rfcs/20210513-pluggable-profiler-for-tensorflow.md

```diff
@@ -0,0 +1,440 @@
+             : plugin_interfaces_(plugin_interfaces) {}
+
+     ~PluggableTracer() override {}
+
+     Status Start() override {
+       for (auto& profiler_interface : plugin_interfaces_) {
+         profiler_interface.DoStart().IgnoreError();
+       }
+       return Status::OK();
+     }
+
+     Status Stop() override {
+       for (auto& plugin_interface : plugin_interfaces_) {
+         plugin_interface.DoStop().IgnoreError();
+       }
+       return Status::OK();
+     }
+     // Unsupported.
+     Status CollectData(RunMetadata* run_metadata) override {
+       for (auto& plugin_interface : plugin_interfaces_) {
+         plugin_interface.DoCollectData(run_metadata).IgnoreError();
+       }
+       return Status::OK();
+     }
+
+     Status CollectData(XSpace* space) override {
+       for (auto& plugin_interface : plugin_interfaces_) {
+         plugin_interface.DoCollectData(space).IgnoreError();
+       }
+       return Status::OK();
+     }
+
+   private:
+     std::vector<PluginTracerInterface> plugin_interfaces_;
+  };
+
+  std::unique_ptr<ProfilerInterface> CreatePluggableTracer(
+    const ProfileOptions& options) {
+    if (options.device_type() != ProfileOptions::PLUGGABLE_DEVICE &&
+        options.device_type() != ProfileOptions::UNSPECIFIED) {
+      return nullptr;
+    }
+    return absl::make_unique<PluggableTracer>(*PluginInterfaceFactory::GetPluginTracerInterfaces());
+  }
+
+
+  auto register_pluggable_tracer_factory = [] {
+    RegisterProfilerFactory(&CreatePluggableTracer);
+    return 0;
+  }();
+
+  ```
+
+#### Plugin
+Define functions that implement `Start()`, `Stop()`, `CollectData()`:
+```c++
+void profiler_start(const TP_Profiler* profiler, TF_Status* status) {
+  // Enable profiler
+  ...
+}
+
+void profiler_stop(const TP_Profiler* profiler, TF_Status* status) {
+  // Disable Profiler
+  ...
+}
+
+void profiler_collect_data_run_metadata(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status) {
+  RunMetadata metadata = get_my_run_metadata(); // Plugin generates RunMetadata based on collected profiler data.
+  *size_in_bytes = metadata.ByteSizeLong(); // get the size of RunMetadata
+  if (buffer == nullptr) {
+    return; // TensorFlow will first get the size of RunMetadata, then allocate the big enough buffer and pass it to plugin for retrieving RunMetadata.
+  }
+  metadata.SerializeToArray(buffer, metadata.ByteSizeLong());
+}
+
+void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status) {
+  Xspace xspace = get_my_xspace(); // Plugin generates Xspace based on collected profiler data.
+  *size_in_bytes = xspace.ByteSizeLong(); // get the size of Xspace
+  if (buffer == nullptr) {
+    return; // TensorFlow will first get the size of Xspace, then allocate the big enough buffer and pass it to plugin for retrieving Xspace.
+  }
+  xspace.SerializeToArray(buffer, xspace.ByteSizeLong());
+}
```

Comment:
1) Need to check the return value of `SerializeToArray()`
2) Need to pass the user's buffer size to SerializeToArray().
3) Need to set the out parameter `status`.


```suggestion
void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status) {
  Xspace xspace = get_my_xspace(); // Plugin generates Xspace based on collected profiler data.
  size_t buffer_size_in_bytes = * size_in_bytes;
  *size_in_bytes = xspace.ByteSizeLong(); // get the size of Xspace
  if (buffer == nullptr) {
    return; // TensorFlow will first get the size of Xspace, then allocate the big enough buffer and pass it to plugin for retrieving Xspace.
  }
  bool success = xspace.SerializeToArray(buffer, buffer_size_in_bytes);
 // TODO: set status to FAILED_PRECONDITION if success is false, OK otherwise.
}
```